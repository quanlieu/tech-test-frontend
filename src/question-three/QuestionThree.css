.three__container {
  display: flex;
  height: calc(100vh - 70px);
}

.three__sideBar {
  background-color: #223048;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  width: 50px;
}

.three__sideBarIcon {
  background-color: #c4c4c4;
  border-radius: 18px;
  height: 36px;
  width: 36px;
  margin: 7px 0;
}

.three__content {
  display: flex;
  flex-direction: column;
  height: calc(100vh - 70px);
  flex-grow: 1;
}

.three__header {
  width: 100%;
}

.three__columns {
  display: flex;
  flex-grow: 1;
  min-height: 0;
  overflow: auto;
}

.three__leftColumn {
  padding: 5px;
  background-color: #e4eef1;
  overflow: scroll;
  width: 34%;
}

.three__rightColumn {
  box-sizing: border-box;
  padding: 5px 5px 0;
  overflow: scroll;
  width: 66%;
}
