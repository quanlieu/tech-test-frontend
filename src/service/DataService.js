import ApolloClient from 'apollo-boost';
import gql from 'graphql-tag';

const graphClient = new ApolloClient({
  uri: 'http://localhost:3500/graphql'
});

export const DataService = {
  getJobsWithSearchTerm: (searchTerm) => {
    return graphClient.query({
      query: gql`
      query ($searchTerm: String){
        jobs(name: $searchTerm) {
          name,
          start,
          end,
          contact {
            id
            name
          }
        }
      }
      `,
      variables: {
        searchTerm: searchTerm
      }
    })
      .then(result => result.data)
      .then(data => data.jobs)
  },

  getJobs: async () => {
    const result = await graphClient.query({
      query: gql`
      query {
        jobs {
          id, name, start, end, location,
        }
      }
      `})
    return result.data.jobs
  },

  getActivities: async () => {
    const result = await graphClient.query({
      query: gql`
      query {
        activities {
          id, name, start, end
        }
      }
      `})
    return result.data.activities
  },

  getResources: async () => {
    const result = await graphClient.query({
      query: gql`
      query {
        resources {
          id, name
        }
      }
      `})
    return result.data.resources
  },

  getJobAllocations: async () => {
    const result = await graphClient.query({
      query: gql`
      query {
        jobAllocations {
          id,
          resource {
            id, name
          },
          job {
            id, name
          }
        }
      }
      `})
    return result.data.jobAllocations
  },

  getActivityAllocations: async () => {
    const result = await graphClient.query({
      query: gql`
      query {
        activityAllocations {
          id,
          resource {
            id, name
          },
          activity {
            id, name
          }
        }
      }
      `})
    return result.data.activityAllocations
  },
}