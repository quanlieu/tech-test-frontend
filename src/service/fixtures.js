export const rawDataSource = {
  resources: [
    { id: '0', name: 'Sam Seaborn' },
    { id: '1', name: 'Donna Moss' },
    { id: '2', name: 'Toby Ziegler' },
  ],
  activities: [
    {
      id: '0',
      name: 'Attend Training',
      start: '2018-09-01T15:00:00Z',
      end: '2018-09-01T17:30:00Z',
    },
    {
      id: '1',
      name: 'Meal Break',
      start: '2018-09-01T12:15:00Z',
      end: '2018-09-01T13:10:00Z',
    },
  ],
  jobs: [
    {
      id: '0',
      name: 'Build a fence',
      location: 'Brisbane',
      start: '2018-09-01T10:00:00Z',
      end: '2018-09-01T11:00:00Z',
    },
    {
      id: '1',
      name: 'Build a shed',
      location: 'Brisbane',
      start: '2018-09-01T10:15:00Z',
      end: '2018-09-01T11:00:00Z',
    },
    {
      id: '2',
      name: 'Shield some wiring',
      location: 'Brisbane',
      start: '2018-09-01T09:00:00Z',
      end: '2018-09-01T13:00:00Z',
    },
    {
      id: '3',
      name: 'Pick up a trailer',
      location: 'Brisbane',
      start: '2018-09-01T13:00:00Z',
      end: '2018-09-01T13:15:00Z',
    },
  ],
  activityAllocations: [
    {
      id: '0',
      resource: { id: '1', name: 'Donna Moss' },
      activity: { id: '1', name: 'Meal Break' },
    },
    {
      id: '1',
      resource: { id: '2', name: 'Toby Ziegler' },
      activity: { id: '1', name: 'Meal Break' },
    },
  ],
  jobAllocations: [
    {
      id: '0',
      resource: { id: '1', name: 'Donna Moss' },
      job: { id: '1', name: 'Build a shed' },
    },
    {
      id: '1',
      resource: { id: '0', name: 'Sam Seaborn' },
      job: { id: '2', name: 'Shield some wiring' },
    },
  ],
}

export const dataSourceResultQuestionTwo = [
  {
    id: '0',
    name: 'Sam Seaborn',
    activities: [],
    jobs: [
      {
        id: '2',
        name: 'Shield some wiring',
        location: 'Brisbane',
        start: '2018-09-01T09:00:00Z',
        end: '2018-09-01T13:00:00Z',
      },
    ],
  },
  {
    id: '1',
    name: 'Donna Moss',
    activities: [
      {
        id: '1',
        name: 'Meal Break',
        start: '2018-09-01T12:15:00Z',
        end: '2018-09-01T13:10:00Z',
      },
    ],
    jobs: [
      {
        id: '1',
        name: 'Build a shed',
        location: 'Brisbane',
        start: '2018-09-01T10:15:00Z',
        end: '2018-09-01T11:00:00Z',
      },
    ],
  },
  {
    id: '2',
    name: 'Toby Ziegler',
    activities: [
      {
        id: '1',
        name: 'Meal Break',
        start: '2018-09-01T12:15:00Z',
        end: '2018-09-01T13:10:00Z',
      },
    ],
    jobs: [],
  },
]

export const dataSourceResultQuestionThree = [
  {
    allocationCount: 0,
    end: '2018-09-01T11:00:00Z',
    id: '0',
    location: 'Brisbane',
    name: 'Build a fence',
    start: '2018-09-01T10:00:00Z',
  },
  {
    allocationCount: 1,
    end: '2018-09-01T11:00:00Z',
    id: '1',
    location: 'Brisbane',
    name: 'Build a shed',
    start: '2018-09-01T10:15:00Z',
  },
  {
    allocationCount: 1,
    end: '2018-09-01T13:00:00Z',
    id: '2',
    location: 'Brisbane',
    name: 'Shield some wiring',
    start: '2018-09-01T09:00:00Z',
  },
  {
    allocationCount: 0,
    end: '2018-09-01T13:15:00Z',
    id: '3',
    location: 'Brisbane',
    name: 'Pick up a trailer',
    start: '2018-09-01T13:00:00Z',
  },
]

export const decoratedDataSourceQuestionTwo = [
  {
    title: 'Sam Seaborn',
    cards: [
      {
        style: { backgroundColor: 'palegreen', color: 'darkgreen' },
        className: 'cardContent',
        description: 'Shield some wiring',
        start: new Date('2018-09-01T09:00:00.000Z'),
        end: new Date('2018-09-01T13:00:00.000Z'),
      },
    ],
  },
  {
    title: 'Donna Moss',
    cards: [
      {
        style: { backgroundColor: 'lightsalmon', color: 'darkred' },
        className: 'cardContent',
        description: 'Meal Break',
        start: new Date('2018-09-01T12:15:00.000Z'),
        end: new Date('2018-09-01T13:10:00.000Z'),
      },
      {
        style: { backgroundColor: 'palegreen', color: 'darkgreen' },
        className: 'cardContent',
        description: 'Build a shed',
        start: new Date('2018-09-01T10:15:00.000Z'),
        end: new Date('2018-09-01T11:00:00.000Z'),
      },
    ],
  },
  {
    title: 'Toby Ziegler',
    cards: [
      {
        style: { backgroundColor: 'lightsalmon', color: 'darkred' },
        className: 'cardContent',
        description: 'Meal Break',
        start: new Date('2018-09-01T12:15:00.000Z'),
        end: new Date('2018-09-01T13:10:00.000Z'),
      },
    ],
  },
]
