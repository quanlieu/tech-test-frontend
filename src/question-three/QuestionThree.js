import React, { useState, useEffect } from 'react'

import { HeaderComponent } from './HeaderComponent'
import { PlaceholderBox } from './PlaceholderBox'
import { JobDetail } from './JobDetail'

import {
  joinDataSourceForQuestionThree,
  fetchDataForQuestionThree,
} from '../service/helpers'

import './QuestionThree.css'

export const QuestionThree = (props) => {
  const [jobs, setJobs] = useState([])
  useEffect(() => {
    const getData = async () => {
      const results = await fetchDataForQuestionThree()
      const dataSource = joinDataSourceForQuestionThree(results)
      setJobs(dataSource)
    }
    getData()
  }, [])

  return (
    <div className="three__container">
      <nav className="three__sideBar">
        <div>
          <div className="three__sideBarIcon">&nbsp;</div>
          <div className="three__sideBarIcon">&nbsp;</div>
          <div className="three__sideBarIcon">&nbsp;</div>
          <div className="three__sideBarIcon">&nbsp;</div>
        </div>
        <div className="three__sideBarIcon">&nbsp;</div>
      </nav>
      <div className="three__content">
        <header className="three__header">
          <HeaderComponent />
        </header>
        <main className="three__columns">
          <div className="three__leftColumn">
            {jobs.map(job => <JobDetail job={job} key={job.id} />)}
          </div>
          <div className="three__rightColumn">
            <PlaceholderBox /><PlaceholderBox /><PlaceholderBox /><PlaceholderBox />
            <PlaceholderBox /><PlaceholderBox /><PlaceholderBox /><PlaceholderBox />
            <PlaceholderBox /><PlaceholderBox /><PlaceholderBox /><PlaceholderBox />
          </div>
        </main>
      </div>
    </div>
  )
}