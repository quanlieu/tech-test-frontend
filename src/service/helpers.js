import { DataService } from "./DataService"

export const arrayToObject = (array, keyField) => array.reduce((obj, item) => {
  obj[item[keyField]] = item
  return obj
}, {})

export const joinDataSourceForQuestionTwo = ({ resources, activities, jobs, activityAllocations, jobAllocations }) => {
  const activitiesObj = arrayToObject(activities, 'id')
  const jobsObj = arrayToObject(jobs, 'id')
  const resourcesObj = arrayToObject(resources.map(r => ({ ...r, activities: [], jobs: [] })), 'id')
  activityAllocations.forEach(ele => {
    resourcesObj[ele.resource.id].activities.push(activitiesObj[ele.activity.id])
  });
  jobAllocations.forEach(ele => {
    resourcesObj[ele.resource.id].jobs.push(jobsObj[ele.job.id])
  });
  return Object.values(resourcesObj)
}

export const fetchDataForQuestionTwo = async () => {
  const {
    getResources, getActivities, getJobs, getActivityAllocations, getJobAllocations,
  } = DataService
  const [
    resources, activities, jobs, activityAllocations, jobAllocations
  ] = await Promise.all([
    getResources(), getActivities(), getJobs(), getActivityAllocations(), getJobAllocations(),
  ])
  return { resources, activities, jobs, activityAllocations, jobAllocations }
}

export const decorateDataForSwimlane = (dataSource) => {
  return dataSource.map(record => ({
    title: record.name,
    cards: [
      ...record.activities.map(activity => ({
        style: {
          backgroundColor: 'lightsalmon',
          color: 'darkred',
        },
        className: 'cardContent',
        description: activity.name,
        start: new Date(activity.start),
        end: new Date(activity.end),
      })),
      ...record.jobs.map(job => ({
        style: {
          backgroundColor: 'palegreen',
          color: 'darkgreen',
        },
        className: 'cardContent',
        description: job.name,
        start: new Date(job.start),
        end: new Date(job.end),
      })),
    ]
  }))
}
