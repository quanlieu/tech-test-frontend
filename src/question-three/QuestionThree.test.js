import React from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import { QuestionThree } from './QuestionThree'
import * as helpers from '../service/helpers'

jest.mock('./JobDetail', () => ({
  JobDetail: () => <div data-testid="job-detail" />,
}))

describe('QuestionThree', () => {
  let useEffect
  const mockUseEffectMount = () => useEffect.mockImplementation((f) => f())

  beforeEach(() => {
    useEffect = jest.spyOn(React, 'useEffect')
    mockUseEffectMount()
    helpers.fetchDataForQuestionThree = jest.fn(() => Promise.resolve({ result: 'ok' }))
    helpers.joinDataSourceForQuestionThree = jest.fn(() => [{ id: 1}, {id: 2}])
  })

  it('renders', async () => {
    expect(render(<QuestionThree />)).toMatchSnapshot()
    await waitFor(() => screen.getAllByTestId('job-detail'))
  })
});
