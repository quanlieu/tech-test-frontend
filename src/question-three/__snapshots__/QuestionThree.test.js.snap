// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionThree renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="three__container"
      >
        <nav
          class="three__sideBar"
        >
          <div>
            <div
              class="three__sideBarIcon"
            >
               
            </div>
            <div
              class="three__sideBarIcon"
            >
               
            </div>
            <div
              class="three__sideBarIcon"
            >
               
            </div>
            <div
              class="three__sideBarIcon"
            >
               
            </div>
          </div>
          <div
            class="three__sideBarIcon"
          >
             
          </div>
        </nav>
        <div
          class="three__content"
        >
          <header
            class="three__header"
          >
            <div
              class="headerComponent"
            >
              Header
            </div>
          </header>
          <main
            class="three__columns"
          >
            <div
              class="three__leftColumn"
            />
            <div
              class="three__rightColumn"
            >
              <div
                class="placeholderBox"
              >
                 
              </div>
              <div
                class="placeholderBox"
              >
                 
              </div>
              <div
                class="placeholderBox"
              >
                 
              </div>
              <div
                class="placeholderBox"
              >
                 
              </div>
              <div
                class="placeholderBox"
              >
                 
              </div>
              <div
                class="placeholderBox"
              >
                 
              </div>
              <div
                class="placeholderBox"
              >
                 
              </div>
              <div
                class="placeholderBox"
              >
                 
              </div>
              <div
                class="placeholderBox"
              >
                 
              </div>
              <div
                class="placeholderBox"
              >
                 
              </div>
              <div
                class="placeholderBox"
              >
                 
              </div>
              <div
                class="placeholderBox"
              >
                 
              </div>
            </div>
          </main>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="three__container"
    >
      <nav
        class="three__sideBar"
      >
        <div>
          <div
            class="three__sideBarIcon"
          >
             
          </div>
          <div
            class="three__sideBarIcon"
          >
             
          </div>
          <div
            class="three__sideBarIcon"
          >
             
          </div>
          <div
            class="three__sideBarIcon"
          >
             
          </div>
        </div>
        <div
          class="three__sideBarIcon"
        >
           
        </div>
      </nav>
      <div
        class="three__content"
      >
        <header
          class="three__header"
        >
          <div
            class="headerComponent"
          >
            Header
          </div>
        </header>
        <main
          class="three__columns"
        >
          <div
            class="three__leftColumn"
          />
          <div
            class="three__rightColumn"
          >
            <div
              class="placeholderBox"
            >
               
            </div>
            <div
              class="placeholderBox"
            >
               
            </div>
            <div
              class="placeholderBox"
            >
               
            </div>
            <div
              class="placeholderBox"
            >
               
            </div>
            <div
              class="placeholderBox"
            >
               
            </div>
            <div
              class="placeholderBox"
            >
               
            </div>
            <div
              class="placeholderBox"
            >
               
            </div>
            <div
              class="placeholderBox"
            >
               
            </div>
            <div
              class="placeholderBox"
            >
               
            </div>
            <div
              class="placeholderBox"
            >
               
            </div>
            <div
              class="placeholderBox"
            >
               
            </div>
            <div
              class="placeholderBox"
            >
               
            </div>
          </div>
        </main>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
